doc/design/logging.md

```diff
@@ -0,0 +1,361 @@
+the compiler will optimize this either to `if (false) iox::log::LogStream(...)`
+and finally completely away or
+`if ((level) <= iox::log::Logger::activeLogLevel()) iox::log::LogStream(...)`.
+The minimal log level check is intended to fully optimize away a log statement
+and the ignore active log level check to always forward the log message to the
+logger, independent of the active log level.
+
+The `self()` invocation is used to create an lvalue reference to the `LogStream`
+object. This eases the implementation of logging support for custom types since
+`IOX_LOG(INFO) << myType;` would require to implement an overload with a rvalue
+`LogStream` reference but `IOX_LOG(INFO) << "#### " << myType;` requires a
+lvalue reference.
+
+The `IOX_LOG` macro is intended for general use and the `IOX_LOG_INTERNAL` for
+special cases when file, line and function are supplied from other places than
+the log macro invocation, like the `Expects` and `Ensures` macros.
+
+Although the macro contains an incomplete if-statement, the `LogStream` object at
+the end makes it safe to use since the compiler will complain if something else
+than a streaming operator or semicolon is used.
+
+#### Behavior before calling initLogger
+
+In order to have log messages before `initLogger` is called, the default logger
+is used with `LogLevel::INFO`. It is up to the implementation of the default
+logger what to do with these messages. For iceoryx the default logger is the
+`ConsoleLogger` (this can be changed via the platform abstraction) which will
+print the log messages to the console.
+
+#### Replacing the logger at runtime
+
+The default console logger can be replaced at runtime to a custom logger which
+can be used to forward the log messages to a logger of another framework.
+
+This is done by deriving from the base `Logger` class and implementing the pure
+virtual `setupNewLogMessage` and `flush` methods. Finally, the derived logger
+must be activated by calling the static `Logger::activeLogger` function and
+passing the new logger as argument to the function. The logger must have a static
+lifetime and should therefore be placed in the data segment.
+
+The call to `iox::log::initLogger` will finalize the option to replace the logger
+at runtime. Further efforts to replace the logger will call the error handler
+with a `MODERATE` error level and a log message to the old and new logger
+
+See also the code example to [create a custom logger](#creating-a-custom-logger).
+
+![logger runtime replacement](../website/images/logger_runtime_replacement.svg)
+
+#### Replacing the default logger for a platform
+
+Each platform must specify the following aliases in the `platform_settings.hpp`
+header.
+
+```cpp
+using LogLevel = pbb::LogLevel;
+using pbb::asStringLiteral;
+
+using Logger = pbb::Logger;
+using DefaultLogger = pbb::ConsoleLogger;
+using TestingLoggerBase = pbb::ConsoleLogger;
+```
+
+With `LogLevel` enum being the log level enum with the values defined in the
```

Comment:
No, it is not intended to have this feature. I'm not sure if it is really valuable or used by many user but it would complicate the code.

Reply:
Ok. Just nice to have and probably easy to add. It would sure add some complexity but I think not much. We can consider it when we need it though.

It was just a question/suggestion.